### PR TITLE
Tools: add ROS 2 launch tests

### PR DESCRIPTION
**Pull request by srmainwaring**
This PR adds two `colcon` packages that enable `colcon` to build and test the `AP_DDS` library in SITL.

There is a brief README in `./Tools/ros2` which includes instructions for running the examples in the `Dockerfile_dev-ros` Docker image.

### Related issues

- https://github.com/ArduPilot/ardupilot/issues/23281
- https://github.com/ArduPilot/ROS2_Tools/issues/1
- https://github.com/ArduPilot/ROS2_Tools/issues/2

## Motivation

The objective is to be able to build and test ArduPilot / ROS 2 integration using `colcon` with commands such as:

```bash
# build
colcon build --packages-select ardupilot_dds_tests

# test
colcon test --packages-select ardupilot_dds_tests

# test results
colcon test-result --all --verbose
```

We'd eventually like to be able to launch a simulation example as

```bash
ros2 launch ardupilot_sitl_gz copter.launch.py
```

A key requirement is that it should be easy for users familiar with ROS to install and run. 

Some non-functional requirements are that the code required to support this be modular and reusable. In particular we'd like to reuse the code for launching in the tests.

## Tasks

- [ ] Ensure tests exit / fail / time out if dependent packages cannot be found.
- [x] Update `ros2.repo` file to point to https://github.com/ArduPilot/ardupilot.git `master`.

## Details

Two packages are added to `./Tools/ros2`.

### `ardupilot_sitl`

A cmake project with a `colcon` package XML that presents the ArduPilot `waf` build as a cmake project, installing targets into the usual workspace folders expected by a ROS 2 `colcon` build. There are a couple of example launch files that will form component parts of the larger test suite for `AP_DDS`. These allow SITL and a non-interactive mavproxy session to be run from the ROS 2 CLI. For example:

```bash
ros2 launch ardupilot_sitl sitl.launch.py
``` 

```bash
ros2 launch ardupilot_sitl mavproxy.launch.py
``` 

The launch files support substitution actions and have launch arguments exposed, so that:

```bash
ros2 launch ardupilot_sitl sitl.launch.py --show-args
Arguments (pass arguments as '<name>:=<value>'):

    'command':
        Run ArduPilot SITL.
        (default: 'arducopter')

    'model':
        Set simulation model.
        (default: 'quad')

    'slave':
        Set the number of JSON slaves.
        (default: '0')

    'sim_address':
        Set address string for simulator.
        (default: '127.0.0.1')

    'speedup':
        Set simulation speedup.
        (default: '1')

    'instance':
        Set instance of SITL (adds 10*instance to all port numbers).
        (default: '0')

    'defaults':
        Set path to defaults file.
        (default: '/Users/rhys/Code/ros2/xrce-dds/ardupilot_ros2_ws/install/ardupilot_sitl/share/ardupilot_sitl/config/default_params/copter.parm')
```

Executables and default parameters are resolved in the launch scripts using the usual ROS 2 package tools.

### `ardupilot_dds_tests`

This package will contain the test suite for the `AP_DDS` library. At present it has a launch file for creating the virtual ports needed for testing SITL:

```bash
ros2 launch ardupilot_dds_tests virtual_ports.launch.py tty0:=./dev/tty0 tty1:=./dev/tty1
[INFO] [launch]: All log files can be found below /Users/rhys/.ros/log/2023-04-07-14-14-46-569511-MacPro2-2.local-9436
[INFO] [launch]: Default logging verbosity is set to INFO
command:      socat
tty0:         ./dev/tty0
tty1:         ./dev/tty1
[INFO] [tty1 -1]: process started with pid [9442]
[tty1 -1] 2023/04/07 14:14:46 socat[9442] N PTY is /dev/ttys024
[tty1 -1] 2023/04/07 14:14:46 socat[9442] N PTY is /dev/ttys025
[tty1 -1] 2023/04/07 14:14:46 socat[9442] N starting data transfer loop with FDs [5,5] and [7,7]
```

There is an example test checking the time message published to the topic /ROS2_Time is received.

```bash
colcon test --packages-select ardupilot_sitl  ardupilot_dds_tests
Starting >>> ardupilot_dds_tests
Starting >>> ardupilot_sitl
Finished <<< ardupilot_sitl [0.54s]                               
--- stderr: ardupilot_dds_tests                     

=============================== warnings summary ===============================
test/test_time_msg_received.py::test_time_msgs_received
test/test_virtual_ports.py::test_read_stdout
  Warning: There is no current event loop

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
---
Finished <<< ardupilot_dds_tests [13.1s]

Summary: 2 packages finished [15.5s]
  1 package had stderr output: ardupilot_dds_tests
```

### CODE REVIEW

File: Tools/ros2/README.md
Comment (Ryanf55) [inline Tools/ros2/README.md:70]: Have you considered also hooking in to the Junit output of colcon test so the tests show up in Gitlab?
https://github.com/marketplace/actions/junit-report-action

File: Tools/ros2/ardupilot_dds_tests/ardupilot_dds_tests/time_listener.py
Comment (Ryanf55) [inline Tools/ros2/ardupilot_dds_tests/ardupilot_dds_tests/time_listener.py:1]: For the python code, have you selected which style you follow, and whether you want to enforce in CI?
ex: `flake8` or `black`

File: Tools/ros2/ardupilot_dds_tests/launch/micro_ros_agent.launch.py
Comment (Ryanf55) [inline Tools/ros2/ardupilot_dds_tests/launch/micro_ros_agent.launch.py:20]: Do you want to validate the refs file exists prior to passing it to the application?

File: Tools/ros2/ardupilot_dds_tests/setup.py
Comment (Ryanf55) [inline Tools/ros2/ardupilot_dds_tests/setup.py:15]: Want to use pathlib?

File: Tools/ros2/ardupilot_dds_tests/test/test_navsat_msg_received.py
Comment (Ryanf55) [inline Tools/ros2/ardupilot_dds_tests/test/test_navsat_msg_received.py:81]: You can set the timeout to 5-10sec  if you want. NavSatFix is usually 5Hz

File: Tools/ros2/ardupilot_sitl/CMakeLists.txt
Comment (Ryanf55) [inline Tools/ros2/ardupilot_sitl/CMakeLists.txt:25]: Do we want to be testing in debug or release? I'd worry if we test in debug, stuff that might be optimized out or behave differently is not caught by CI. Perhaps set `--debug` only if CMAKE_BUILD_TYPE=DEBUG?

File: Tools/ros2/ardupilot_sitl/CMakeLists.txt
Comment (Ryanf55) [inline Tools/ros2/ardupilot_sitl/CMakeLists.txt:43]: Please remove commented code

File: Tools/ros2/ardupilot_sitl/CMakeLists.txt
Comment (Ryanf55) [inline Tools/ros2/ardupilot_sitl/CMakeLists.txt:114]: Can you add in ament_lint_common, or something similar to enfore style/syntax in XML, python, and cmake code?

File: Tools/ros2/ardupilot_sitl/launch/sitl.launch.py
Comment (Ryanf55) [inline Tools/ros2/ardupilot_sitl/launch/sitl.launch.py:70]:
```suggestion
    for label in range(len(uart_labels)+1):
```

Avoid magic numbers please

File: Tools/ros2/ardupilot_sitl/package.xml
Comment (Ryanf55) [inline Tools/ros2/ardupilot_sitl/package.xml:8]: Want to fill in your name?

File: Tools/ros2/ardupilot_sitl/tests/test_a.py
Comment (Ryanf55) [inline Tools/ros2/ardupilot_sitl/tests/test_a.py:1]: Example test can be removed

File: Tools/ros2/ros2.repos
Comment (Ryanf55) [inline Tools/ros2/ros2.repos:4]: Will this be migrated before PR merges?

**Review by Ryanf55 — CHANGES_REQUESTED** [review]
Great work so far. Going to go ahead and do some testing to validate. A few minor points here and there.

File: Tools/ros2/ardupilot_dds_tests/ardupilot_dds_tests/time_listener.py
Comment (srmainwaring) [reply Tools/ros2/ardupilot_dds_tests/ardupilot_dds_tests/time_listener.py:1]: I've been using `black` formatter for work with Gazebo, but ROS appears to use `flake8`. I've hovered between the two with this project (`blake7`?) but think we should align with ROS standard and use `flake8`. I'll set up the linter / formatter and should probably reinstate the testing the comes with the default Python package created by `colcon`.

File: Tools/ros2/README.md
Comment (srmainwaring) [reply Tools/ros2/README.md:70]: Not looked into this. Sounds like a good idea for the GHW.

File: Tools/ros2/ardupilot_dds_tests/launch/micro_ros_agent.launch.py
Comment (srmainwaring) [reply Tools/ros2/ardupilot_dds_tests/launch/micro_ros_agent.launch.py:20]: I haven't added validation to the any arguments. I guess that's a policy we might want to decide - whether to put validation in what's effectively a pass through / translation layer, add it over the top, or leave it to the underlying node to handle. Whatever we decide it should prob apply to all arguments so it's consistent.

File: Tools/ros2/ardupilot_dds_tests/setup.py
Comment (srmainwaring) [reply Tools/ros2/ardupilot_dds_tests/setup.py:15]: Yes - could do. Pathlib being the modern approach right? By and large I followed exiting examples from the ROS source and the Humble tutorials. `launch` is circa 2015 - so prob a bit out of date in terms of Python flavour.

File: Tools/ros2/ardupilot_sitl/CMakeLists.txt
Comment (srmainwaring) [reply Tools/ros2/ardupilot_sitl/CMakeLists.txt:25]: Yes - release. There are a few macOS specific items I need to swap out. The SITL release build on macOS has been know to seg fault in the past which is why these are here. It should be release every where though. Will change.

File: Tools/ros2/ros2.repos
Comment (srmainwaring) [reply Tools/ros2/ros2.repos:4]: Yes. Although it will mean it can't be tested until the PR is merged. There's a bootstrap problem. We need the PR in to retrieve the cmake enabled build from master. Was thinking that for this and similar we might immediately follow this PR with a patch to set repo versions correct? Or I could just change it and you trust me it works.

File: Tools/ros2/ardupilot_sitl/CMakeLists.txt
Comment (srmainwaring) [reply Tools/ros2/ardupilot_sitl/CMakeLists.txt]: Changed in 77504b1bc6

File: Tools/ros2/ardupilot_sitl/launch/sitl.launch.py
Comment (srmainwaring) [reply Tools/ros2/ardupilot_sitl/launch/sitl.launch.py:70]: Introduced variable `max_serial_ports`.

Changed in f60fd749f7

File: Tools/ros2/ardupilot_sitl/package.xml
Comment (srmainwaring) [reply Tools/ros2/ardupilot_sitl/package.xml]: Changed in ced7893b13

No escape now!

File: Tools/ros2/README.md
Comment (Ryanf55) [reply Tools/ros2/README.md:70]: You can cover it in a follow up PR if you like. Looking at test reports in GH UI can be nicer than scrolling through long logs.

File: Tools/ros2/ardupilot_dds_tests/ardupilot_dds_tests/time_listener.py
Comment (Ryanf55) [reply Tools/ros2/ardupilot_dds_tests/ardupilot_dds_tests/time_listener.py:1]: Yea, only problem with `flake8` is it doesn't appear the tool can format your code for you. The ROS community has considered switching, and even thought it would be a good idea for new packages, but `ament_black` hasn't yet been added to the standard set of linters. If the ArduPilot devs are ok, I'd prefer to use black since it's automatic.

File: Tools/ros2/ros2.repos
Comment (Ryanf55) [reply Tools/ros2/ros2.repos]: I'm fine with a patch

File: Tools/ros2/README.md
Comment (srmainwaring) [reply Tools/ros2/README.md:70]: Yes, let's do that.

File: Tools/ros2/ardupilot_sitl/CMakeLists.txt
Comment (srmainwaring) [reply Tools/ros2/ardupilot_sitl/CMakeLists.txt]: I've enabled `ament_lint_auto` in `ardupilot_sitl` to see how it looks. That package is now compliant with all linters **and** formatted with `black`. There is a way to override the default ament flake8 config. The overrides are in the `.flake8` config file. There are only two changes needed - line length and double quotes. There may be more as we add code, but it seems to be relatively pain free.

**Comment by srmainwaring** [comment]
## Update

- Moved launch files in `ardupilot_dds_tests` to `ardupilot_sitl`. Launch files being used downstream should not be in a `test` project.
- Renamed the launch file for `sitl_dds`.
- Update README and refactor tests to accommodate launch file changes.

**Comment by tridge** [comment]
@Ryanf55 are you ok to approve this?

**Review by Ryanf55 — APPROVED** [review]
LGTM. Minor updates can be done after it bootstraps.

File: Tools/ros2/README.md
Comment (srmainwaring) [reply Tools/ros2/README.md:53]: Defer to #23549

File: Tools/ros2/README.md
Comment (srmainwaring) [reply Tools/ros2/README.md:62]: Defer to #23549

**Comment by davidbuzz** [comment]
it does seem a bit weird to me to use colcon to execute cmake to execute waf ...  when all three of them are build tools with similar capabilities.

**Comment by Ryanf55** [comment]
> it does seem a bit weird to me to use colcon to execute cmake to execute waf ... when all three of them are build tools with similar capabilities.

If ArduPilot supported CMake, then it would only be two, but we're just using what's available. It's a bit of an odd marriage, but that's the best idea we had for this initial version.

**Comment by srmainwaring** [comment]
@davidbuzz it seems a bit strange at first, but it's a mechanism to translate ArduPilot's core repo into the ROS world without making any intrusive changes.

Here's some of the reasoning behind it: `colcon` is more of a package management system / build workflow coordinator than a build system like Python `setuptools` or `cmake`. ROS 2 `ament`, which extends `colcon`, dictates the structure of ROS 2 packages so the tooling can locate and find libraries, binaries and other resources. In order for those to work we need to make ArduPilot look and feel like a set of ROS 2 packages and follow their rules. They have no rules for `waf`, so the package has to be either `cmake` or Python. The Python packages have a number of limitations - very stringent directory structure and no symlink installs - so `cmake` was the better option. We also want to make ArduPilot appear as a number of ROS packages - one for core library, one for tests, perhaps others to come. This precludes setting the package structure at the ArduPilot root (which is also more intrusive). Finally, if we want to eventually distribute built packages as .debs or the widows equivalents, so ROS users can download pre-built versions of ArduPilot for ROS (which would make setting up new users that already have a working ROS much easier) then I think we need to go all in with the ROS build / install process. The resulting installation from the colcon build is completely divorced from the ArduPilot source structure and will run independently of that. I don't believe that is true for things like sim_vehicle.py and autotools which rely on things being in certain places in the ArduPilot source tree.

**Pull request merged**
